worlds/kdl3/regions.py

```diff
@@ -119,46 +120,50 @@ def generate_rooms(world: "KDL3World", level_regions: Dict[int, Region]):
-            try:
-                if levels[level][stage] is None:
-                    stage_candidates = [candidate for candidate in possible_stages
-                                        if (enforce_world and candidate in default_levels[level])
-                                        or (enforce_pattern and ((candidate - 1) & 0x00FFFF) % 6 == stage)
-                                        or (enforce_pattern == enforce_world)
-                                        ]
-                    new_stage = generate_valid_level(world, level, stage, stage_candidates, levels[level])
+def generate_valid_levels(world: "KDL3World", shuffle_mode: int) -> Dict[int, List[int]]:
+    if shuffle_mode:
+        levels: Dict[int, List[Optional[int]]] = {
+            1: [None] * 7,
+            2: [None] * 7,
+            3: [None] * 7,
+            4: [None] * 7,
+            5: [None] * 7,
+        }
+
+        possible_stages = [default_levels[level][stage] for level in default_levels for stage in range(6)]
+        if world.multiworld.plando_connections[world.player]:
+            for connection in world.multiworld.plando_connections[world.player]:
+                try:
+                    entrance_world, entrance_stage = connection.entrance.rsplit(" ", 1)
+                    stage_world, stage_stage = connection.exit.rsplit(" ", 1)
+                    new_stage = default_levels[location_name.level_names[stage_world.strip()]][int(stage_stage) - 1]
+                    levels[location_name.level_names[entrance_world.strip()]][int(entrance_stage) - 1] = new_stage
                     possible_stages.remove(new_stage)
-                    levels[level][stage] = new_stage
-            except Exception:
-                raise Exception(f"Failed to find valid stage for {level}-{stage}. Remaining Stages:{possible_stages}")
 
+                except Exception:
+                    raise Exception(
+                        f"Invalid connection: {connection.entrance} =>"
+                        f" {connection.exit} for player {world.player} ({world.player_name})")
+
+        for level in range(1, 6):
+            for stage in range(6):
+                # Randomize bosses separately
+                try:
+                    if levels[level][stage] is None:
+                        stage_candidates = [candidate for candidate in possible_stages
+                                            if (shuffle_mode == 1 and candidate in default_levels[level])
+                                            or (shuffle_mode == 2 and ((candidate - 1) & 0x00FFFF) % 6 == stage)
+                                            or (shuffle_mode == 3)
+                                            ]
+                        new_stage = generate_valid_level(world, level, stage, stage_candidates, levels[level])
+                        possible_stages.remove(new_stage)
+                        levels[level][stage] = new_stage
+                except Exception:
```

Comment:
Does this have to be a try-except? It seems like you know exactly what the exception means, wouldn't it be simpler to check the related condition instead? To avoid throwing two exceptions?